### ISSUE TITLE
[SPIKE]:  TorexCore - the Exchange Mechanism

### ISSUE THREAD

**Issue by tokdaniel** [superfluid-finance/superfluid-stories#453]
## Overview
TOREX (Twap-ORacle EXchange) is the Superfluid-only exchange mechanism with streams. It one of the main aggregation routes of AverageX. TOREX aims to help any average trader joes setting up a hussle-free and consistent stream of investment into the assets they LOVE.

**What?** 
 - Contract implementing stream-in token A -> distribute-out token B
 - (optional) A factory that deploys markets
 
**How?**
- Leverage [Uniswap-v3 TWAP](https://blog.uniswap.org/uniswap-v3-oracles) (Time Weighed Average Price) and its spot price oracle.
- A _liquidity mover_ provides the necessary liquidity for distributions in a timely manner.

To be investigated: What on-chain constraints for an acceptable price point is required & sufficient for reliable incentives?
Is the TWAP price enough or do we also need to take into account the spot price? Other input parameters needed?

**Why?**
  - This is the hearth and soul of AverageX.
  - To have a concrete technical solution on how to do the stream-swapping between assets.

## Acceptance criteria
- [ ] A TDD to outline the implementation details for a production version of the TOREX contracts
 - [ ] A working/testable POC which can `stream-swap` from one asset to another on a test network.
 - [ ] (Good to have) A factory to deploy contracts for assetpairs.

## Links/Resources
*   https://www.notion.so/superfluidhq/TOREX-17979bb24d0d4901a355293083788a44?pvs=4
*   https://www.blog.uniswap.org/uniswap-v3-oracles 
*   https://www.docs.uniswap.org/concepts/protocol/oracle 

## Deadline
17th. nov
        
## Readiness criteria
- [x] Descriptive title
- [x] What?
- [x] Why?
- [x] How?
- [x] Acceptance criteria complying with Definition of Done
- [x] Estimate/Story points
- [x] Deadline: If not a part of an epic
- [x] Priority: Reflected as order on the bord, walking the board from the top and right
- [x] Assignee
- [x] Labels
- [x] Link to epic: If belonging to an epic
- [x] A task is always a separate ticket/issue
- [x] UI/UX Design (optional)
- [x] Blockers/Prerequisites (optional): Use Zenhub blocked by/blocking feature
